File: .claude/skills/sast-scan/rules/semgrep/python/tests/test_taint_rules.py
import subprocess
import pickle
import pathlib
import requests
import urllib.request


def taint_sql(cursor, request):
    user_id = request.args.get("id")
    sql = "SELECT * FROM users WHERE id = " + user_id
    # ruleid: python.security.taint-sql-injection
    cursor.execute(sql)

    # ok: python.security.taint-sql-injection
    cursor.execute("SELECT * FROM users WHERE id = %s", (1,))


def taint_command(request):
    cmd = request.form.get("cmd")
    # ruleid: python.security.taint-command-injection
    subprocess.run(cmd, shell=True)

    # ok: python.security.taint-command-injection
    subprocess.run(["ls", "-la"])


def taint_eval(request):
    expr = request.GET.get("expr")
    # ruleid: python.security.taint-eval
    eval(expr)

    # ok: python.security.taint-eval
    eval("1 + 1")


def taint_path(request):
    path = request.args.get("file")
    # ruleid: python.security.taint-path-traversal
    open(path, "r")

    # ok: python.security.taint-path-traversal
    open("/etc/app/config.json", "r")


def taint_ssrf(request):
    url = request.form.get("url")
    # ruleid: python.security.taint-ssrf
    requests.get(url)

    # ok: python.security.taint-ssrf
    requests.get("https://api.example.com/health")


def taint_pickle(request):
    data = request.args.get("blob")
    # ruleid: python.security.taint-pickle
    pickle.loads(data.encode())

File: .claude/skills/sast-scan/rules/semgrep/javascript/tests/test_taint_rules.js
const { exec, execSync } = require("child_process");
const fs = require("fs");

function taintSql(req, db) {
  const userId = req.query.id;
  const sql = "SELECT * FROM users WHERE id = " + userId;
  // ruleid: javascript.security.taint-sql-injection
  db.query(sql);

  // ok: javascript.security.taint-sql-injection
  db.query("SELECT * FROM users WHERE id = ?", [1]);
}

function taintCommand(req) {
  const cmd = req.body.cmd;
  // ruleid: javascript.security.taint-command-injection
  exec(cmd);

  // ok: javascript.security.taint-command-injection
  exec("ls -la");
}

function taintXss(req, res) {
  const msg = req.query.msg;
  // ruleid: javascript.security.taint-xss
  res.send(msg);

  // ok: javascript.security.taint-xss
  res.send("static");
}

function taintPath(req) {
  const userPath = req.params.file;
  // ruleid: javascript.security.taint-path-traversal
  fs.readFile(userPath, "utf8", () => {});

  // ok: javascript.security.taint-path-traversal
  fs.readFile("/var/app/data.json", "utf8", () => {});
}

function taintSsrf(req) {
  const url = req.query.url;
  // ruleid: javascript.security.taint-ssrf
  fetch(url);

  // ok: javascript.security.taint-ssrf
  fetch("https://api.example.com/health");
}

function taintRedirect(req, res) {
  const target = req.query.next;
  // ruleid: javascript.security.taint-open-redirect
  res.redirect(target);

  // ok: javascript.security.taint-open-redirect
  res.redirect("/dashboard");
}
